components/expand-collapse/expand-collapse.js

```diff
@@ -0,0 +1,132 @@
+import { css, html, LitElement } from 'lit-element/lit-element.js';
+import ResizeObserver from 'resize-observer-polyfill/dist/ResizeObserver.es.js';
+import { styleMap } from 'lit-html/directives/style-map.js';
+
+const reduceMotion = matchMedia('(prefers-reduced-motion: reduce)').matches;
+
+const states = {
+	COLLAPSE: 'collapse', // in the process of collapsing or fully collapsed
+	EXPANDING: 'expanding', // in the process of expanding
+	EXPANDED: 'expanded', // fully expanded
+};
+
+class ExpandCollapse extends LitElement {
+
+	static get properties() {
+		return {
+			expanded: { type: Boolean, reflect: true },
+			_height: { type: Number },
+			_state: { type: String }
+		};
+	}
+
+	static get styles() {
+		return css`
+			:host {
+				display: block;
+			}
+
+			:host([hidden]) {
+				display: none;
+			}
+
+			.d2l-expand-collapse-container {
+				height: 0px;
+				overflow: hidden;
+				transition: height 400ms cubic-bezier(0, 0.7, 0.5, 1);
+			}
+
+			.d2l-expand-collapse-container[data-state="expanded"] {
+				overflow: visible;
+			}
+
+			/* prevent margin colapse on slotted children */
+			.d2l-expand-collapse-content:before,
+			.d2l-expand-collapse-content:after {
+				content: ' ';
+				display: table;
+			}
+
+			@media (prefers-reduced-motion: reduce) {
+				.d2l-expand-collapse-container {
+					transition: none;
+				}
+			}
+		`;
+	}
+
+	constructor() {
+		super();
+		this._onContentResize = this._onContentResize.bind(this);
+		this.expanded = false;
+		this._state = states.COLLAPSE;
+	}
+
+	get expanded() {
+		return this._expanded;
+	}
+
+	set expanded(val) {
+		const oldVal = this._expanded;
+		if (oldVal !== val) {
+			this._expanded = val;
+			this.requestUpdate('expanded', oldVal);
+			this._expandedChanged(val);
+		}
+	}
+
+	firstUpdated() {
+		super.firstUpdated();
+
+		const content = this.shadowRoot.querySelector('.d2l-expand-collapse-content');
+		this._resizeObserver = new ResizeObserver(this._onContentResize);
+		this._resizeObserver.observe(content);
```

Comment:
If the observer ends up being buggy, we could set the height to `auto` once it's fully expanded. i.e. only calculate the height when it initially goes from collapsed -> expanded to do the animation. Then when the animation finishes,  remove the fixed height.

Reply:
I think this is preferred (reverting to `auto`) - currently, if the browser is resized while open, the `height` property is updated.  I think it would be better to let the browser handle this natively.  

If we _do_ need the observer, we should try to make use of its [entries](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserverEntry) to avoid making measurements during such times where the user is resizing, etc. The observer provides this because it's such a common (costly) pattern.

Reply:
Reverting to auto is turning out to be surprisingly painful because transitions won't run to / from `auto. ` This makes the close transition flakey because the height needs to be computed then set to a specific value before collapsing which requires waiting for a timeout for the recalc to occur, then setting the height to 0. This works the majority of the time but if `expanded` is changed while the timeout is in progress then we run into race conditions. Still working to see if I can find a way around this.